lib/pipelining/event.hpp

```diff
@@ -0,0 +1,208 @@
+/**
+ * @file event.h
+ * @author Kudryashov Ilya (kudriashov.it@phystech.edu)
+ * @brief Event class
+ * @version 0.1
+ * @date 2024-03-20
+ *
+ * @copyright Copyright (c) 2024
+ *
+ */
+
+#pragma once
+
+#include <functional>
+#include <set>
+#include <stdexcept>
+
+/**
+ * @brief Event class. Can be triggered to notify its subscribers and deliver a
+ * payload to them.
+ *
+ * @tparam Ts payload contents
+ */
+template <class... Ts>
+struct Event final {
+    /**
+     * @brief Event listener
+     *
+     */
+    struct Listener {
+        explicit Listener(const std::function<void(Ts...)>& action)
+            : action_(action){};
+
+        Listener() : Listener([](Ts...) {}) {}
+
+        Listener(const Listener& other)
+            : action_(other.action_), event_(other.event_) {
+            if (event_) event_->subscribe(*this);
+        }
+
+        Listener& operator=(const Listener& other) {
+            if (&other == this) return;
+
+            if (event_) event_->unsubscribe(*this);
+
+            action_ = other.action_;
+            event_ = other.event_;
+
+            if (event_) event_->subscribe(*this);
+
+            return *this;
+        }
+
+        Listener(Listener&& other)
+            : action_(other.action_), event_(other.event_) {
+            if (event_) event_->subscribe(*this);
+            if (other.event_) other.event_->unsubscribe(other);
+        }
+
+        Listener& operator=(Listener&& other) {
+            if (event_) event_->unsubscribe(*this);
+
+            action_ = other.action_;
+            event_ = other.event_;
+
+            if (event_) event_->subscribe(*this);
+
+            if (other.event_) other.event_->unsubscribe(other);
+
+            return *this;
+        }
+
+        ~Listener() {
+            if (event_) event_->unsubscribe(*this);
+        }
+
+        typename Event<Ts...>::Listener& operator=(Listener& other);
```

Comment:
using listener_t = typename  Event<Ts...>::Listener?

Reply:
Not sure about that...
It would only be used once. And public `using listener_t = ...;` inside the class literally caller `Listener` looks a bit weird.
Why would an object contain a typedef for itself?
```C++
Event<...>::Listener::listener_t::listener_t::listener_t::listener_t::...::listener_t listener;
```
```C++
Event<...>::Listener listener;
```

Reply:
Мб я не понимаю, что такое typename? Зачем он тут